src/main.rs

```diff
@@ -0,0 +1,223 @@
+
+    pub const DEFAULT_DEST_ADDR: &str = "127.0.0.1:9443";
+
+    pub fn parse_enclave_addr(address: &str) -> Result<EnclaveAddr> {
+        Ok(SocketAddr::from_str(address).context("error parsing destination address")?)
+    }
+
+    pub async fn connect_to_enclave(address: EnclaveAddr) -> Result<EnclaveStream> {
+        Ok(TcpStream::connect(address).await?)
+    }
+
+    pub async fn shutdown_enclave_stream(stream: &mut EnclaveStream) {
+        stream.shutdown().await.ok();
+    }
+}
+
+/// Relays TCP connections from IPV4/IPV6 to VSOCK.
+#[derive(Parser, Debug)]
+#[command(version)]
+struct Cli {
+    /// Buffer size to use when reading/writing data between peers
+    #[arg(long, default_value_t = 8192)]
+    buffer_size: usize,
+
+    /// IPV4/IPV6 address/port to listen on
+    #[arg(short = 's', long, default_value = "0.0.0.0:8443")]
+    source_address: String,
+
+    /// VSOCK address/port to connect to
+    #[arg(short = 'l', long, default_value = DEFAULT_DEST_ADDR)]
+    destination_address: String,
+
+    /// Maximum amount of allowed concurrent connections
+    #[arg(short = 'c', long, default_value_t = 1250)]
+    max_concurrent_connections: usize,
+}
+
+struct RelayTask {
+    src_conn: TcpStream,
+    dest_conn: EnclaveStream,
+    src_rx_bytes: BytesMut,
+    dest_rx_bytes: BytesMut,
+}
+
+impl RelayTask {
+    pub async fn new(
+        src_conn: TcpStream,
+        dest_addr: EnclaveAddr,
+        buffer_size: usize,
+    ) -> Result<Self> {
+        let dest_conn = connect_to_enclave(dest_addr).await?;
+        Ok(Self {
+            src_conn,
+            dest_conn,
+            src_rx_bytes: BytesMut::with_capacity(buffer_size),
+            dest_rx_bytes: BytesMut::with_capacity(buffer_size),
+        })
+    }
+
+    async fn shutdown(&mut self) {
+        self.src_conn.shutdown().await.ok();
+        shutdown_enclave_stream(&mut self.dest_conn).await;
+    }
+
+    async fn handle_rx_result(&mut self, rx_result: io::Result<usize>) -> Result<bool> {
+        match rx_result {
+            Ok(bytes_read) => {
+                // If bytes_read == 0, assume connection has terminated
+                if bytes_read == 0 {
+                    self.shutdown().await;
+                    return Ok(false);
+                }
+            }
+            Err(e) => {
+                self.shutdown().await;
+                return Err(e.into());
+            }
+        }
+        Ok(true)
+    }
+
+    async fn handle_dest_conn_rx(&mut self, rx_result: io::Result<usize>) -> Result<bool> {
+        if !self.handle_rx_result(rx_result).await? {
+            debug!("recv empty buf from dest connection, quitting comm");
+            return Ok(false);
+        }
+        self.src_conn.write_buf(&mut self.dest_rx_bytes).await?;
+        Ok(true)
+    }
+
+    async fn handle_src_conn_rx(&mut self, rx_result: io::Result<usize>) -> Result<bool> {
+        if !self.handle_rx_result(rx_result).await? {
```

Comment:
is this guaranteed to be a non-zero buffer? From what I can tell here the src_connection comes from the requesting client to the STAR server so could an empty request get through to this code point and be used to shutdown the connection between the proxy shim and the enclave?

Reply:
Yes, if zero bytes are read, then the peer terminated the connection. Even if that were not the case, only the current connection would be shutdown; all other connections would remain intact.

Reply:
Cool no concerns with that then